### ISSUE TITLE
Segmentation fault in converter

### ISSUE THREAD

**Issue by wvanasten** [Azure/kubelogin#107]
I have been having trouble getting the config converter to work. It always segfaults, with what looks like a null pointer and afterwards it silently "succeeds", because of the lock file.

```
Platform: Archlinux (5.15.55-1-lts)
Go version:  go version go1.18.4 linux/amd64
Distribution: https://aur.archlinux.org/packages/azure-kubelogin
```

I had been noticing this when using the program, but this morning I noticed it also fails when running the converter test (Which luckily seems to notice the lock file and no longer silently succeeds on the second attempt).

Any help/pointing in the right direction would be greatly appreciated.

The stacktrace:

```
unexpected fault address 0x0
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x80 addr=0x0 pc=0x55c1d8ae89df]

goroutine 10 [running]:
runtime.throw({0x55c1d9935374?, 0x30?})
        runtime/panic.go:992 +0x71 fp=0xc0002ca0b8 sp=0xc0002ca088 pc=0x55c1d8ab6e51
runtime.sigpanic()
        runtime/signal_unix.go:825 +0x305 fp=0xc0002ca108 sp=0xc0002ca0b8 pc=0x55c1d8acdb65
aeshashbody()
        runtime/asm_amd64.s:1343 +0x39f fp=0xc0002ca110 sp=0xc0002ca108 pc=0x55c1d8ae89df
runtime.mapiternext(0xc000628d80)
        runtime/map.go:934 +0x2cb fp=0xc0002ca180 sp=0xc0002ca110 pc=0x55c1d8a8f7ab
runtime.mapiterinit(0x1?, 0x7?, 0x1?)
        runtime/map.go:861 +0x228 fp=0xc0002ca1a0 sp=0xc0002ca180 pc=0x55c1d8a8f488
reflect.mapiterinit(0x55c1d8a944a6?, 0x7ff2230f5a00?, 0x55c1d8b6cb47?)
        runtime/map.go:1373 +0x19 fp=0xc0002ca1c8 sp=0xc0002ca1a0 pc=0x55c1d8ae47d9
github.com/modern-go/reflect2.(*UnsafeMapType).UnsafeIterate(...)
        github.com/modern-go/reflect2@v1.0.1/unsafe_map.go:112
github.com/json-iterator/go.(*sortKeysMapEncoder).Encode(0xc0002227b0, 0xc000111120, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_map.go:291 +0x225 fp=0xc0002ca338 sp=0xc0002ca1c8 pc=0x55c1d8f3b285
github.com/json-iterator/go.(*placeholderEncoder).Encode(0x55c1d9c80c00?, 0x1?, 0xc00010f038?)
        github.com/json-iterator/go@v1.1.10/reflect.go:332 +0x22 fp=0xc0002ca360 sp=0xc0002ca338 pc=0x55c1d8f340a2
github.com/json-iterator/go.(*structFieldEncoder).Encode(0xc000222a20, 0x55c1d9b7e76d?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:110 +0x56 fp=0xc0002ca3d8 sp=0xc0002ca360 pc=0x55c1d8f489d6
github.com/json-iterator/go.(*structEncoder).Encode(0xc000222a80, 0x1500?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:158 +0x765 fp=0xc0002ca4c0 sp=0xc0002ca3d8 pc=0x55c1d8f493e5
github.com/json-iterator/go.(*OptionalEncoder).Encode(0xc00010f020?, 0xc000121310?, 0xc0002ca550?)
        github.com/json-iterator/go@v1.1.10/reflect_optional.go:70 +0xa4 fp=0xc0002ca510 sp=0xc0002ca4c0 pc=0x55c1d8f405e4
github.com/json-iterator/go.(*placeholderEncoder).Encode(0x55c1d9c80c00?, 0xc000628c01?, 0xc00010f038?)
        github.com/json-iterator/go@v1.1.10/reflect.go:332 +0x22 fp=0xc0002ca538 sp=0xc0002ca510 pc=0x55c1d8f340a2
github.com/json-iterator/go.(*structFieldEncoder).Encode(0xc00011a330, 0x55c1d9bb07dc?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:110 +0x56 fp=0xc0002ca5b0 sp=0xc0002ca538 pc=0x55c1d8f489d6
github.com/json-iterator/go.(*structEncoder).Encode(0xc00011a3f0, 0xc0001214f0?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:158 +0x765 fp=0xc0002ca698 sp=0xc0002ca5b0 pc=0x55c1d8f493e5
github.com/json-iterator/go.(*placeholderEncoder).Encode(0x55c1d9c80c00?, 0x55c1d8f48101?, 0xc00010f038?)
        github.com/json-iterator/go@v1.1.10/reflect.go:332 +0x22 fp=0xc0002ca6c0 sp=0xc0002ca698 pc=0x55c1d8f340a2
github.com/json-iterator/go.(*structFieldEncoder).Encode(0xc00011a660, 0x55c1d9b7e743?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:110 +0x56 fp=0xc0002ca738 sp=0xc0002ca6c0 pc=0x55c1d8f489d6
github.com/json-iterator/go.(*structEncoder).Encode(0xc00011a6c0, 0x55c1d9c30980?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:158 +0x765 fp=0xc0002ca820 sp=0xc0002ca738 pc=0x55c1d8f493e5
github.com/json-iterator/go.(*sliceEncoder).Encode(0xc000111638, 0xc000240478, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_slice.go:38 +0x2bd fp=0xc0002ca8d8 sp=0xc0002ca820 pc=0x55c1d8f4103d
github.com/json-iterator/go.(*structFieldEncoder).Encode(0xc00011b2c0, 0x55c1d9b83f4d?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:110 +0x56 fp=0xc0002ca950 sp=0xc0002ca8d8 pc=0x55c1d8f489d6
github.com/json-iterator/go.(*structEncoder).Encode(0xc00011b3b0, 0x0?, 0xc00010f020)
        github.com/json-iterator/go@v1.1.10/reflect_struct_encoder.go:158 +0x765 fp=0xc0002caa38 sp=0xc0002ca950 pc=0x55c1d8f493e5
github.com/json-iterator/go.(*OptionalEncoder).Encode(0xc0001ef720?, 0x0?, 0x0?)
        github.com/json-iterator/go@v1.1.10/reflect_optional.go:70 +0xa4 fp=0xc0002caa88 sp=0xc0002caa38 pc=0x55c1d8f405e4
github.com/json-iterator/go.(*onePtrEncoder).Encode(0xc0002271d0, 0xc000240420, 0xc0004f4150?)
        github.com/json-iterator/go@v1.1.10/reflect.go:219 +0x82 fp=0xc0002caac0 sp=0xc0002caa88 pc=0x55c1d8f33662
github.com/json-iterator/go.(*Stream).WriteVal(0xc00010f020, {0x55c1d9ce6460, 0xc000240420})
        github.com/json-iterator/go@v1.1.10/reflect.go:98 +0x158 fp=0xc0002cab30 sp=0xc0002caac0 pc=0x55c1d8f32978
github.com/json-iterator/go.(*frozenConfig).Marshal(0xc0001ef720, {0x55c1d9ce6460, 0xc000240420})
        github.com/json-iterator/go@v1.1.10/config.go:299 +0xc9 fp=0xc0002cabc8 sp=0xc0002cab30 pc=0x55c1d8f29309
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).doEncode(0x55c1d9b650f2?, {0x55c1d9e0a6e8?, 0xc000240420?}, {0x55c1d9e02b08, 0xc000806a20})
        k8s.io/apimachinery@v0.21.3/pkg/runtime/serializer/json/json.go:306 +0x6d fp=0xc0002cac60 sp=0xc0002cabc8 pc=0x55c1d90803ad
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Encode(0xc0000aafa0, {0x55c1d9e0a6e8, 0xc000240420}, {0x55c1d9e02b08, 0xc000806a20})
        k8s.io/apimachinery@v0.21.3/pkg/runtime/serializer/json/json.go:301 +0xfc fp=0xc0002cacc0 sp=0xc0002cac60 pc=0x55c1d90802dc
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).doEncode(0xc0001efa40, {0x55c1d9e0a698?, 0xc00010ef60}, {0x55c1d9e02b08, 0xc000806a20})
        k8s.io/apimachinery@v0.21.3/pkg/runtime/serializer/versioning/versioning.go:244 +0x8fa fp=0xc0002cb018 sp=0xc0002cacc0 pc=0x55c1d908ef9a
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode(0xc0001efa40, {0x55c1d9e0a698, 0xc00010ef60}, {0x55c1d9e02b08, 0xc000806a20})
        k8s.io/apimachinery@v0.21.3/pkg/runtime/serializer/versioning/versioning.go:184 +0x106 fp=0xc0002cb078 sp=0xc0002cb018 pc=0x55c1d908e646
k8s.io/apimachinery/pkg/runtime.Encode({0x7ff2203fb298, 0xc0001efa40}, {0x55c1d9e0a698, 0xc00010ef60})
        k8s.io/apimachinery@v0.21.3/pkg/runtime/codec.go:50 +0x64 fp=0xc0002cb0b8 sp=0xc0002cb078 pc=0x55c1d8f883e4
k8s.io/client-go/tools/clientcmd.Write(...)
        k8s.io/client-go@v0.21.3/tools/clientcmd/loader.go:466
k8s.io/client-go/tools/clientcmd.WriteToFile({{0x0, 0x0}, {0x0, 0x0}, {0x0, 0xc0007b78f0}, 0xc0007b7920, 0xc0007b7950, 0xc0007b7980, {0x55c1d99345b8, ...}, ...}, ...)
        k8s.io/client-go@v0.21.3/tools/clientcmd/loader.go:419 +0xa8 fp=0xc0002cb130 sp=0xc0002cb0b8 pc=0x55c1d90ba048
k8s.io/client-go/tools/clientcmd.writeCurrentContext({0x55c1d9e199b8, 0xc000204a00}, {0x55c1d99345b8, 0x3})
        k8s.io/client-go@v0.21.3/tools/clientcmd/config.go:415 +0x1a8 fp=0xc0002cb318 sp=0xc0002cb130 pc=0x55c1d90b70c8
k8s.io/client-go/tools/clientcmd.ModifyConfig({0x55c1d9e199b8, 0xc000204a00}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0xc0003af500}, 0xc0003af530, 0xc0003af560, ...}, ...)
        k8s.io/client-go@v0.21.3/tools/clientcmd/config.go:197 +0x198 fp=0xc0002cba40 sp=0xc0002cb318 pc=0x55c1d90b59b8
github.com/Azure/kubelogin/pkg/converter.Convert({0xc00029a300, {0x55c1d9e19008, 0xc0003af5f0}, {{0x55c1d99394d5, 0xa}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}})
        github.com/Azure/kubelogin/pkg/converter/convert.go:317 +0x2a47 fp=0xc0002cbd18 sp=0xc0002cba40 pc=0x55c1d992a247
github.com/Azure/kubelogin/pkg/converter.TestConvert.func1(0xc00020ba00)
        github.com/Azure/kubelogin/pkg/converter/convert_test.go:930 +0x479 fp=0xc0002cbf70 sp=0xc0002cbd18 pc=0x55c1d99319b9
testing.tRunner(0xc00020ba00, 0xc00007af00)
        testing/testing.go:1439 +0x102 fp=0xc0002cbfc0 sp=0xc0002cbf70 pc=0x55c1d8ba5f22
testing.(*T).Run.func1()
        testing/testing.go:1486 +0x2a fp=0xc0002cbfe0 sp=0xc0002cbfc0 pc=0x55c1d8ba6dca
runtime.goexit()
        runtime/asm_amd64.s:1571 +0x1 fp=0xc0002cbfe8 sp=0xc0002cbfe0 pc=0x55c1d8aeb661
created by testing.(*T).Run
        testing/testing.go:1486 +0x35f

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00020b380, {0x55c1d993a1db?, 0x55c1d8ba5905?}, 0x55c1d9dfbae8)
        testing/testing.go:1487 +0x37a
testing.runTests.func1(0xc00020b380?)
        testing/testing.go:1839 +0x6e
testing.tRunner(0xc00020b380, 0xc000507cd8)
        testing/testing.go:1439 +0x102
testing.runTests(0xc000465d60?, {0x55c1da770360, 0x2, 0x2}, {0x7ff249bf0108?, 0x40?, 0x55c1da827ce0?})
        testing/testing.go:1837 +0x457
testing.(*M).Run(0xc000465d60)
        testing/testing.go:1719 +0x5d9
main.main()
        _testmain.go:49 +0x1aa

goroutine 6 [chan receive]:
k8s.io/klog.(*loggingT).flushDaemon(0x0?)
        k8s.io/klog@v1.0.0/klog.go:1010 +0x6a
created by k8s.io/klog.init.0
        k8s.io/klog@v1.0.0/klog.go:411 +0xef

goroutine 7 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x0?)
        k8s.io/klog/v2@v2.8.0/klog.go:1164 +0x6a
created by k8s.io/klog/v2.init.0
        k8s.io/klog/v2@v2.8.0/klog.go:418 +0xf6

goroutine 9 [chan receive]:
testing.(*T).Run(0xc00020b520, {0x55c1d994499c?, 0x55c1d994462c?}, 0xc00007af00)
        testing/testing.go:1487 +0x37a
github.com/Azure/kubelogin/pkg/converter.TestConvert(0xc00020b520?)
        github.com/Azure/kubelogin/pkg/converter/convert_test.go:904 +0x6f7c
testing.tRunner(0xc00020b520, 0x55c1d9dfbae8)
        testing/testing.go:1439 +0x102
created by testing.(*T).Run
        testing/testing.go:1486 +0x35f
FAIL    github.com/Azure/kubelogin/pkg/converter        0.024s
```

**Comment by weinong** [comment]
you have to use go 1.17 for now. I haven't merged the PR using 1.18

**Comment by weinong** [comment]
@wvanasten can you give #96 a try?

**Comment by wvanasten** [comment]
@weinong Yes, it works with #97.

As an aside: It appears that (one of?) the tests actually modifies the config file located in `~/.kube`.

**Comment by weinong** [comment]
fixed by #97 and #113
